contracts/Donations.sol

```diff
@@ -0,0 +1,199 @@
+        emit TTLUpdated(_ttl);
+    }
+
+    /**
+     * Transfers the donated funds in the currency @param _token to the charity with the id @param _destinationId.
+     *
+     * @param _destinationId ID of the charity.
+     * @param _token Currency to transfer the funds from.
+     * @param _to Address of the charity.
+     */
+    function donate(
+        uint256 _destinationId,
+        IERC20 _token,
+        address _to
+    ) external onlyRole(WORKER_ROLE) {
+        require(
+            transferableAmounts[_token][_destinationId] != 0,
+            "Donations: nothing to donate"
+        );
+
+        uint256 amount = transferableAmounts[_token][_destinationId];
+        transferableAmounts[_token][_destinationId] = 0;
+
+        _token.safeTransfer(_to, amount);
+
+        emit DonationsSent(_destinationId, _token, _to, amount);
+    }
+
+    /**
+     * Mints an NFT for every donation in @param _params.
+     * The @param _txHash and @param _batchNr uniquely identify a group of donations.
+     *
+     * @param _txHash The hash of the transaction where the yield for these donations was claimed.
+     * @param _batchNr When there are too many donations in a claim, we break them into batches to not reach the gas limit.
+     * @param _params Donation params.
+     */
+    function mint(
+        bytes32 _txHash,
+        uint256 _batchNr,
+        DonationParams[] calldata _params
+    ) external onlyRole(WORKER_ROLE) {
+        bytes32 groupId = keccak256(abi.encodePacked(_txHash, _batchNr));
+
+        require(
+            !processedDonationsGroups[groupId],
+            "Donations: already processed"
+        );
+
+        uint64 expiry = _getBlockTimestamp() + ttl;
+        uint256 length = _params.length;
+
+        for (uint256 i = 0; i < length; i++) {
+            uint256 _metadataId = metadataId.current();
+            metadataId.increment();
+
+            metadata[_metadataId] = Metadata({
+                destinationId: _params[i].destinationId,
+                token: _params[i].token,
+                expiry: expiry,
+                amount: _params[i].amount
+            });
+
+            _mint(_params[i].owner, _metadataId);
+
+            emit DonationMinted(
+                _metadataId,
+                _params[i].destinationId,
+                groupId,
+                _params[i].token,
+                expiry,
+                _params[i].amount,
+                _params[i].owner
+            );
+        }
+
+        processedDonationsGroups[groupId] = true;
+    }
+
+    /**
+     * Burns the NFT and sets the amount donated to be transferred to the charity.
+     *
+     * @param _id ID of the NFT.
+     */
+    function burn(uint256 _id) external {
+        bool isOwner = ownerOf(_id) == _msgSender();
+        bool isWorker = hasRole(WORKER_ROLE, _msgSender());
+        bool expired = metadata[_id].expiry <= _getBlockTimestamp();
+
+        require(isOwner || (isWorker && expired), "Donations: not allowed");
```

Comment:
when expired, only worker can burn NFT? is there any issue if anybody calls after expired?

Reply:
Discussed on Slack and updated 👍